cranelift/codegen/src/isa/unwind/winx64.rs

```diff
@@ -219,30 +224,49 @@ impl UnwindInfo {
     ) -> CodegenResult<Self> {
         use crate::isa::unwind::input::UnwindCode as InputUnwindCode;
 
+        let word_size: u32 = unwind.word_size.into();
         let mut unwind_codes = Vec::new();
         for c in unwind.prologue_unwind_codes.iter() {
             match c {
-                InputUnwindCode::SaveRegister { offset, reg } => {
-                    unwind_codes.push(UnwindCode::PushRegister {
-                        offset: ensure_unwind_offset(*offset)?,
-                        reg: MR::map(*reg),
-                    });
-                }
-                InputUnwindCode::StackAlloc { offset, size } => {
-                    unwind_codes.push(UnwindCode::StackAlloc {
-                        offset: ensure_unwind_offset(*offset)?,
-                        size: *size,
-                    });
-                }
-                InputUnwindCode::SaveXmmRegister {
+                InputUnwindCode::SaveRegister {
                     offset,
                     reg,
                     stack_offset,
                 } => {
-                    unwind_codes.push(UnwindCode::SaveXmm {
+                    let reg = MR::map(*reg);
+                    let offset = ensure_unwind_offset(*offset)?;
+                    match reg {
+                        MappedRegister::Int(reg) => {
+                            let push_reg_sequence = if let Some(UnwindCode::StackAlloc {
```

Comment:
Could you help me understand what this is trying to do since no similar check existed prior when saving integer registers?

Reply:
push usually consists of "sp += 8" and "save register" operation, and for win64 it is trying to recover/merge these two operations into `UnwindCode::PushRegister`

Reply:
I'm not sure I follow.  The unwind operation on Windows x64 is called "push nonvolatile register" for integer register saves.  The unwind code doesn't need to know what the stack pointer value was or the operand size because they're always word-sized (i.e. 8 bytes).

Therefore it shouldn't need to look at the stack allocation operation at all to record an unwind operation.

Additionally, the integer register saves are performed in the prologue *before* the stack pointer adjustment, so I don't get how this is supposed to work as the last unwind code would never be `StackAlloc`.

A typical x64 integer register saving prologue with the associated unwind codes:

```asm
push rbp          ; UnwindOperation::PushNonvolatileRegister
mov rbp, rsp
push r12          ; UnwindOperation::PushNonvolatileRegister
push r13          ; UnwindOperation::PushNonvolatileRegister
sub rsp, 0x20     ; UnwindOperation::SmallStackAlloc
```

Reply:
I'm pretty sure this block should just be:

```rust
unwind_codes.push(UnwindCode::PushRegister { offset, reg });
```

Right?

Reply:
Above example is correct. I think you missed the change in cranelift/codegen/src/isa/x86/unwind.rs  near `Push64` which now has `input::UnwindCode::StackAlloc` in addition to `::StoreRegister`.

FWIW Old backend is using only push contstructs, new backend produces more "sub sp, 16 / mov rxx, 0(rsp) / mov ryy, 8(rsp)" - like stuff in x64 and aarch64 backends, so eventually we will start doing more tricky pushes.

Reply:
I'm a little concerned these changes didn't cause test failures for the Windows unwind tests 😕

Reply:
The unwind information output shall not be affected at least for win64; at least I tried not to affect the final result.

Reply:
I'm getting it now.  Could we add a comment to this block explaining why it is expecting a `StackAlloc` preceding the `SaveRegister` operation to simulate a "push" and thus the push operation should be `UWOP_PUSH_NONVOL`?